### PR TITLE
GEL: add expressions to query editor (w/ feature flag)

### PR DESCRIPTION
**Pull request by ryantxu**
This PR moves work from GEL app to the standard frontend query editor.

TODO (but not in this PR)
* expressions can only run on things that have backend support.  We need to either:
  * show an error if you have them configured
  * don't allow configuring things without backend support

NOTE: this PR depends on #20088 to pass all hidden queries to the DataSource

### CODE REVIEW

File: public/app/plugins/datasource/expr/ExpressionDatasource.ts
Comment (ryantxu) [inline public/app/plugins/datasource/expr/ExpressionDatasource.ts]: This is a hack... but I'm not sure a better option.  I want to reuse the QueryEditor interface, BUT do not want ExpressionDatasource to be a real datasource.  other ideas?

File: public/app/plugins/datasource/expr/ExpressionDatasource.ts
Comment (ryantxu) [inline public/app/plugins/datasource/expr/ExpressionDatasource.ts]: @kylebrandt @marefr -- note this uses `/api/ds` we should change that to whatever matches the new endpoint

**Comment by ryantxu** [comment]
@torkelo -- can I get quick feedback on this?  In particular what do you think about the not-really-a-datasource datasource?

I don't think it should be defined in `public/app/plugins/datasource/expr/` -- but I am not sure where makes more sense.

File: public/app/plugins/datasource/expr/ExpressionDatasource.ts
Comment (marefr) [reply public/app/plugins/datasource/expr/ExpressionDatasource.ts]: We can use whatever we find suitable. The one that exists now is a v2/copy of tsdb query and we probably want to use something else :+1:

File: public/app/plugins/datasource/expr/ExpressionDatasource.ts
Comment (marefr) [reply public/app/plugins/datasource/expr/ExpressionDatasource.ts]: I leave this up to someone that knows front end better.

File: public/app/plugins/datasource/expr/ExpressionQueryEditor.tsx
Comment (marefr) [inline public/app/plugins/datasource/expr/ExpressionQueryEditor.tsx]: We may want to move all of the reducer,  downsampling and upsampling types to the backend eventually, but no blocker for now. Idea would be that the transform plugin can be queried for meta data about the functions it supports?

**Review by marefr — COMMENTED** [review]
Looks like a great start. Haven't had the time to test it out yet

File: public/app/plugins/datasource/expr/ExpressionQueryEditor.tsx
Comment (ryantxu) [reply public/app/plugins/datasource/expr/ExpressionQueryEditor.tsx]: This is something we should investigate -- it really depends what kind of operations we need and how "custom" the UI needs to get.  it could be something that generates the UI automatically -- or just needs to stay in sync with the frontend.  This would not be a plugin that we "support" extending -- we may have different options in enterprise though

File: public/app/plugins/datasource/expr/ExpressionDatasource.ts
Comment (torkelo) [reply public/app/plugins/datasource/expr/ExpressionDatasource.ts]: should this not use datasourceRequest() function (otherwise query inspector does not work)

File: pkg/api/metrics.go
Comment (torkelo) [inline pkg/api/metrics.go]: Feels like this function QueryMetricsV2 and QueryMetrics could need some refactoring, there is some duplication between the two and this new function is a bit long and messy, for example the duplicate !expr conditions can maybe be reduced to one if building the queries loop was extracted to a function.

File: public/app/plugins/datasource/expr/ExpressionDatasource.ts
Comment (torkelo) [reply public/app/plugins/datasource/expr/ExpressionDatasource.ts]: can't think of any better ideas either.

**Review by torkelo — COMMENTED** [review]
Looked better than I expected :)  I could not find any major things I would want changed before merge except maybe the backend ap go code could need some refactoring.

File: public/app/features/expressions/ExpressionDatasource.ts
Comment (ryantxu) [inline public/app/features/expressions/ExpressionDatasource.ts]: @torkelo -- I changed this to use `datasourceRequest`, that is not exposed on the runtime getBackendSrv().

NOTE: I tried to change the method types to return Promise, but it broke a bunch of datasources so I added the cast here

Comment (kylebrandt) [reply public/app/features/expressions/ExpressionDatasource.ts]: Doesn't seem to be working for me with GEL:

```
Panel data error: TypeError: Cannot convert undefined or null to object
    at Function.values (<anonymous>)
    at gelResponseToDataFrames (default~app.777340c294f1a9edf1c6.js:47686)
    at default~app.777340c294f1a9edf1c6.js:47435
    at processQueue (angular~app.777340c294f1a9edf1c6.js:17178)
    at angular~app.777340c294f1a9edf1c6.js:17226
    at Scope.$digest (angular~app.777340c294f1a9edf1c6.js:18361)
    at Scope.$apply (angular~app.777340c294f1a9edf1c6.js:18658)
    at angular~app.777340c294f1a9edf1c6.js:18961
    at completeOutstandingRequest (angular~app.777340c294f1a9edf1c6.js:6437)
    at angular~app.777340c294f1a9edf1c6.js:6716
```

I do see results from the network call in Chrome inspector:

```
{"results":{"":{"refId":"","series":null,"tables":null,"dataframes":["QVJST1cxAAC8AQAAEAAAAAAACgAOAAwACwAEAAoAAAAUAAAAAAAAAQMACgAMAAAACAAEAAoAAAAIAAAAWAAAAAIAAAAoAAAABAAAAND+//8IAAAADAAAAAEAAABBAAAABQAAAHJlZklkAAAA8P7//wgAAAAUAAAACAAAAEEtc2VyaWVzAAAAAAQAAABuYW1lAAAAAAIAAACcAAAABAAAAH7///8UAAAAZAAAAGQAAAAAAAMBZAAAAAIAAAAsAAAABAAAAEj///8IAAAAEAAAAAYAAABudW1iZXIAAAQAAAB0eXBlAAAAAGz///8IAAAAEAAAAAUAAAB2YWx1ZQAAAAQAAABuYW1lAAAAAAAAAABi////AAACAAUAAAB2YWx1ZQASABgAFAATABIADAAAAAgABAASAAAAFAAAAGwAAAB0AAAAAAAKAXQAAAACAAAANAAAAAQAAADc////CAAAABAAAAAEAAAAdGltZQAAAAAEAAAAdHlwZQAAAAAIAAwACAAEAAgAAAAIAAAAEAAAAAQAAAB0aW1lAAAAAAQAAABuYW1lAAAAAAAAAAAAAAYACAAGAAYAAAAAAAMABAAAAHRpbWUAAAAAvAAAABQAAAAAAAAADAAWABQAEwAMAAQADAAAALAAAAAAAAAAFAAAAAAAAAMDAAoAGAAMAAgABAAKAAAAFAAAAFgAAAALAAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAWAAAAAAAAABYAAAAAAAAAAAAAAAAAAAAWAAAAAAAAABYAAAAAAAAAAAAAAACAAAACwAAAAAAAAAAAAAAAAAAAAsAAAAAAAAAAAAAAAAAAAAAAAAAAJhC+m1r0hUA8Inye2vSFQBI0eqJa9IVAKAY45dr0hUA+F/bpWvSFQBQp9Oza9IVAKjuy8Fr0hUAADbEz2vSFQBYfbzda9IVALDEtOtr0hUACAyt+WvSFQAAAAAAAABAAAAAAAAA8D8AAAAAAADwPwAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQAAAAAAAAABAAAAAAAAA8D8AAAAAAADwPwAAAAAAAAAAAAAAAAAAAAAQAAAADAAUABIADAAIAAQADAAAABAAAAAsAAAAOAAAAAAAAwABAAAAyAEAAAAAAADAAAAAAAAAALAAAAAAAAAAAAAAAAAAAAAAAAoADAAAAAgABAAKAAAACAAAAFgAAAACAAAAKAAAAAQAAADQ/v//CAAAAAwAAAABAAAAQQAAAAUAAAByZWZJZAAAAPD+//8IAAAAFAAAAAgAAABBLXNlcmllcwAAAAAEAAAAbmFtZQAAAAACAAAAnAAAAAQAAAB+////FAAAAGQAAABkAAAAAAADAWQAAAACAAAALAAAAAQAAABI////CAAAABAAAAAGAAAAbnVtYmVyAAAEAAAAdHlwZQAAAABs////CAAAABAAAAAFAAAAdmFsdWUAAAAEAAAAbmFtZQAAAAAAAAAAYv///wAAAgAFAAAAdmFsdWUAEgAYABQAEwASAAwAAAAIAAQAEgAAABQAAABsAAAAdAAAAAAACgF0AAAAAgAAADQAAAAEAAAA3P///wgAAAAQAAAABAAAAHRpbWUAAAAABAAAAHR5cGUAAAAACAAMAAgABAAIAAAACAAAABAAAAAEAAAAdGltZQAAAAAEAAAAbmFtZQAAAAAAAAAAAAAGAAgABgAGAAAAAAADAAQAAAB0aW1lAAAAAOgBAABBUlJPVzE=","QVJST1cxAACsAQAAEAAAAAAACgAOAAwACwAEAAoAAAAUAAAAAAAAAQMACgAMAAAACAAEAAoAAAAIAAAAUAAAAAIAAAAoAAAABAAAAOD+//8IAAAADAAAAAEAAABCAAAABQAAAHJlZklkAAAAAP///wgAAAAMAAAAAAAAAAAAAAAEAAAAbmFtZQAAAAACAAAAlAAAAAQAAACG////FAAAAGAAAABgAAAAAAADAWAAAAACAAAALAAAAAQAAABQ////CAAAABAAAAAGAAAAbnVtYmVyAAAEAAAAdHlwZQAAAAB0////CAAAAAwAAAAAAAAAAAAAAAQAAABuYW1lAAAAAAAAAABm////AAACAAAAAAAAABIAGAAUABMAEgAMAAAACAAEABIAAAAUAAAAbAAAAHQAAAAAAAoBdAAAAAIAAAA0AAAABAAAANz///8IAAAAEAAAAAQAAAB0aW1lAAAAAAQAAAB0eXBlAAAAAAgADAAIAAQACAAAAAgAAAAQAAAABAAAAFRpbWUAAAAABAAAAG5hbWUAAAAAAAAAAAAABgAIAAYABgAAAAAAAwAEAAAAVGltZQAAAAC8AAAAFAAAAAAAAAAMABYAFAATAAwABAAMAAAAsAAAAAAAAAAUAAAAAAAAAwMACgAYAAwACAAEAAoAAAAUAAAAWAAAAAsAAAAAAAAAAAAAAAQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABYAAAAAAAAAFgAAAAAAAAAAAAAAAAAAABYAAAAAAAAAFgAAAAAAAAAAAAAAAIAAAALAAAAAAAAAAAAAAAAAAAACwAAAAAAAAAAAAAAAAAAAAAAAAAAmEL6bWvSFQDwifJ7a9IVAEjR6olr0hUAoBjjl2vSFQD4X9ula9IVAFCn07Nr0hUAqO7LwWvSFQAANsTPa9IVAFh9vN1r0hUAsMS062vSFQAIDK35a9IVAAAAAAAAEEAAAAAAAAAIQAAAAAAAAAhAAAAAAAAAAEAAAAAAAAAAQAAAAAAAABBAAAAAAAAAEEAAAAAAAAAIQAAAAAAAAAhAAAAAAAAAAEAAAAAAAAAAQBAAAAAMABQAEgAMAAgABAAMAAAAEAAAACwAAAA4AAAAAAADAAEAAAC4AQAAAAAAAMAAAAAAAAAAsAAAAAAAAAAAAAAAAAAAAAAACgAMAAAACAAEAAoAAAAIAAAAUAAAAAIAAAAoAAAABAAAAOD+//8IAAAADAAAAAEAAABCAAAABQAAAHJlZklkAAAAAP///wgAAAAMAAAAAAAAAAAAAAAEAAAAbmFtZQAAAAACAAAAlAAAAAQAAACG////FAAAAGAAAABgAAAAAAADAWAAAAACAAAALAAAAAQAAABQ////CAAAABAAAAAGAAAAbnVtYmVyAAAEAAAAdHlwZQAAAAB0////CAAAAAwAAAAAAAAAAAAAAAQAAABuYW1lAAAAAAAAAABm////AAACAAAAAAAAABIAGAAUABMAEgAMAAAACAAEABIAAAAUAAAAbAAAAHQAAAAAAAoBdAAAAAIAAAA0AAAABAAAANz///8IAAAAEAAAAAQAAAB0aW1lAAAAAAQAAAB0eXBlAAAAAAgADAAIAAQACAAAAAgAAAAQAAAABAAAAFRpbWUAAAAABAAAAG5hbWUAAAAAAAAAAAAABgAIAAYABgAAAAAAAwAEAAAAVGltZQAAAADYAQAAQVJST1cx"]}}}```

Comment (kylebrandt) [reply public/app/features/expressions/ExpressionDatasource.ts]: Seems like the promise is being rejected?

File: pkg/api/metrics.go
Comment (kylebrandt) [reply pkg/api/metrics.go]: Reduced it some. I'm not sure about sharing code between the two endpoints since I think there might be some differences in Q2? Currently every query object with a request has datasource info, but they will all be executed based on the first datasource in the list of queries. Which is confusing. I guess it is fine because generally as things requests are grouped by datasource anyways (with the exception of gel). 

So in short I am still not clear on how we want the V2 endpoint to work yet.

Comment (torkelo) [reply pkg/api/metrics.go]: yea, that is fine for now 👍

**Review by torkelo — APPROVED** [review]
LGTM, no blockers!

**Pull request merged**
